### PR TITLE
[MachineLICM] Hoist COPY instruction only when user can be hoisted

### PR DESCRIPTION
**Pull request by michaelselehov**
befa925acac8fd6a9266e introduced preliminary hoisting of COPY instructions when the user of the COPY is inside the same loop. That optimization appeared to be too aggressive and hoisted too many COPY's greatly increasing register pressure causing performance regressions for AMDGPU target.

This is intended to fix the regression by hoisting COPY instruction only if either:
 - User of COPY can be hoisted (other args are invariant) 
 or
 - Hoisting COPY doesn't bring high register pressure

### CODE REVIEW

**Review by Asher8118 — APPROVED** [review]
The changes make sense to me. @davemgreen might want to take a look as well.

**Comment by davemgreen** [comment]
Do you have a test demonstrating the problem you are running into? We usually prefer that code is accompanied by a test case to make sure we don't accidentally break it again in the future.

**Comment by michaelselehov** [comment]
> Do you have a test demonstrating the problem you are running into? We usually prefer that code is accompanied by a test case to make sure we don't accidentally break it again in the future.

I have an IR derived from the benchmark which is pretty big. I reduced it down to the point when only one spill reload still appears but yet it's 685 lines big. I'll see if I can make anything prettier.

**Comment by michaelselehov** [comment]
> Do you have a test demonstrating the problem you are running into? We usually prefer that code is accompanied by a test case to make sure we don't accidentally break it again in the future.

I managed to reduce the test as much as I could. It's still 80+ lines of IR but the check is simple: there should be no spill reload in the loop.

File: llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll [hunk 1]
```diff
@@ -0,0 +1,94 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -O2 -vectorize-loops -vectorize-slp -amdgpu-early-inline-all=true -amdgpu-function-calls=false < %s | FileCheck %s
+target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8:9-p9:192:256:256:32"
+target triple = "amdgcn-amd-amdhsa"
+
+define amdgpu_kernel void @_ZN6Kokkos4ImplL32hip_parallel_launch_local_memoryINS0_11ParallelForIN4Test7HexGradINS_3HIPEdfEENS_11RangePolicyIJS5_EEES5_EELj1024ELj1EEEvT_(ptr %.sroa.1.0.copyload, ptr %0, ptr %1, ptr %2, ptr %3, ptr %4, ptr %5, ptr %6, ptr %7, ptr %8, ptr %9, ptr %10, ptr %11, ptr %12, ptr %13, ptr %14, ptr %15, ptr %16, ptr %17, ptr %18, ptr %19, ptr %20, ptr %21, ptr %22, ptr %23, ptr %24, ptr %25, ptr %26, ptr %27, ptr %28, ptr %29, ptr %30, ptr %31, ptr %32, ptr %33, double %34, double %35, double %36, float %37, float %38, float %39, float %40, ptr %41) {
```
[the remark this answers is not in the document]
Comment (michaelselehov) [reply llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll]: Good point.

[the remark this answers is not in the document]
Comment (Rin18) [reply llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll]: Thanks for updating the test.

File: llvm/lib/CodeGen/MachineLICM.cpp [hunk 2]
```diff
@@ -1264,13 +1264,24 @@ bool MachineLICMBase::IsProfitableToHoist(MachineInstr &MI,
 
   // If we have a COPY with other uses in the loop, hoist to allow the users to
   // also be hoisted.
+  Register defReg;
```
Comment (davemgreen) [inline llvm/lib/CodeGen/MachineLICM.cpp:1267]: defReg -> DefReg

File: llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll [hunk 3]
```diff
@@ -0,0 +1,94 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
```
Comment (davemgreen) [inline llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll:1]: The check lines do not look auto-generated, but they might be a bit fragile as-is. Would it be possible to auto-generate them? Or is that too large?
An explanation of what is going wrong in the test (some-such thing should not be hoisted and spilled?) might be useful in the long-run too.

**Review by davemgreen — APPROVED** [review]
Thanks. The code looks OK to me

File: llvm/lib/CodeGen/MachineLICM.cpp (hunk 2 above)
Comment (michaelselehov) [reply llvm/lib/CodeGen/MachineLICM.cpp]: Done.

File: llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll (hunk 3 above)
Comment (michaelselehov) [reply llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll]: I started with autogenerated test, hence the artefact in the NOTE.
But I liked my way of checking much more than the autogeneration. The latter is going to fail from the every little sneeze in the compiler that changes sequence of instructions of register names. Instead my way of checking checks exactly what I want: spill reload in the loop. And when the test fails the error is very clear to read.

**Comment by michaelselehov** [comment]
@Rin18 @davemgreen if you're fine with the change, would you please do me a favor and merge it? Thanks.

File: llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll [hunk 4]
```diff
@@ -0,0 +1,94 @@
+; NOTE: There must be no spill reload inside the loop starting with LBB0_1:
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -O2 -vectorize-loops -vectorize-slp -amdgpu-early-inline-all=true -amdgpu-function-calls=false < %s | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll:2]: -amdgpu-early-inline-all=true -amdgpu-function-calls=false shouldn't do anything with llc. Don't think -vectorize-loops -vectorize-slp do either, and -O2 is the default

Comment (michaelselehov) [reply llvm/test/CodeGen/AMDGPU/copy-hoist-no-spills.ll]: Thanks. Fixed.

**Comment by michaelselehov** [comment]
There seems to be a weird error with windows/flang test that must not be related to my PR. Waiting for a fix.

**Comment by davemgreen** [comment]
The windows tests look like they should be unrelated. Lets get this in - thanks.

**Comment by michaelselehov** [comment]
> The windows tests look like they should be unrelated. Lets get this in - thanks.

Thank you for merging!

**Pull request merged**